### PR TITLE
Graduate DocValuesNumbersQuery from lucene/sandbox to newSlowSetQuery()

### PR DESCRIPTION
**Pull request by rmuir**
Clean up this query a bit, and move it around to support:

* NumericDocValuesField.newSlowSetQuery()
* SortedNumericDocValuesField.newSlowSetQuery()

This complements the existing docvalues-based range queries, with a set query.

Later we can hook this into IntField/LongField/FloatField/DoubleField via IndexOrDocValuesQuery.

In general cleanup was not a big deal, involves:
* fix code to use e.g. DocValues.isCacheable rather than assuming docvalues can't be updated
* implement optimized codepath for single-valued fields
* in general, try to be consistent with SortedNumericDocValuesRangeQuery as much as possible

Relates to #12028

### CODE REVIEW

**Comment by rmuir** [comment]
intended as followups:
* look into PointRangeQuery and implement necessary estimation for IndexOrDocValuesQuery to do the right thing
* Add newSetQuery() to IntField/LongField/DoubleField/FloatField, that uses IndexOrDocValuesQuery(PointRangeQuery, ThisQuery)

**Comment by rmuir** [comment]
Here's my benchmarks with attached java program: [NumSetBenchmark.java.txt](https://github.com/apache/lucene/files/10419558/NumSetBenchmark.java.txt)
* `main` uses `IntPoint.newSetQuery` on main branch
* `patch` uses `IntField.newSetQuery` on this branch.

The purpose was to run different batches of "hard" queries to look for performance regressions (not using numeric IDs, but terms of various density intersecting integer sets of various density). The reported time in ms. is the time it takes to run the batch

I don't see any problems:

| Query Set  | main (IndexSearcher.count) | patch (IndexSearcher.count) | main (IndexSearcher.search) | patch (IndexSearcher.search)
| ------------- | ------------- | ------------- | ------------- | ------------- |
| BIG_BIG | 14.43ms  | 11.30ms  | 11.75ms  | 5.98ms  |
| MEDIUM_BIG  | 16.45ms  | 6.25ms  | 17.08ms  | 5.66ms  |
| SMALL_BIG  | 17.54ms  | 2.00ms  | 18.43ms  | 2.52ms  |
| BIG_MEDIUM | 5.50ms  | 4.54ms  | 5.90ms | 5.00ms  |
| MEDIUM_MEDIUM | 6.39ms  | 3.70ms  | 7.13ms  | 4.70ms  |
| SMALL_MEDIUM | 6.64ms  | 1.43ms  | 6.98ms  | 1.70ms  |

**Comment by rmuir** [comment]
the benchmark above uses queries such as `"la|21,22,23",    // 2226 hits`

in this case we form a boolean query of TermQuery:"la" AND admin2code in (21,22,23). The admin2 codes are typically county level in most countries and each one these numbers match many documents: e.g. 100,000+

I ran the benchmark on full geonames (11M+ docs)

File: lucene/core/src/java/org/apache/lucene/document/NumericDocValuesField.java [hunk 1]
```diff
@@ -97,6 +97,27 @@ SortedNumericDocValues getValues(LeafReader reader, String field) throws IOExcep
     };
   }
 
+  /**
+   * Create a query matching any of the specified values.
+   *
+   * <p><b>NOTE</b>: Such queries cannot efficiently advance to the next match, which makes them
+   * slow if they are not ANDed with a selective query. As a consequence, they are best used wrapped
+   * in an {@link IndexOrDocValuesQuery}, alongside a set query that executes on points, such as
+   * {@link LongPoint#newSetQuery}.
```
Comment (jpountz) [inline lucene/core/src/java/org/apache/lucene/document/NumericDocValuesField.java:106]: Maybe link to LongField and other similar fields?

**Review by jpountz — APPROVED** [review]
The change and benchmark results look great!

File: lucene/core/src/java/org/apache/lucene/document/NumericDocValuesField.java (hunk 1 above)
Comment (rmuir) [reply lucene/core/src/java/org/apache/lucene/document/NumericDocValuesField.java:106]: oh, duh, thank you. will fix.

**Review by jpountz — APPROVED** [review]

**Pull request merged**